### DISCUSSION TITLE
Vertical Workspaces

### DISCUSSION THREAD

**Discussion by michaelhthomas** [PaperWM-community/PaperWM#6] [Ideas]
I and many others are using PaperWM in conjunction with the Vertical Workspaces plugin for Gnome 40. As the vertical workspace layout is pretty integral to the PaperWM workflow, is this something we should add as a module to PaperWM proper?

**Comment by paradajz** [comment]
I would very much like if Vertical overview extension would actually get merged into this fork. Conceptually, using horizontal workspaces with PaperWM just feels... off. Even in real life, if you had two slightly longer scrolls, you would probably arrange them vertically when using/reading/examining/other them as opposed to horizontal direction. Also some screenshots in README suggest vertical workspaces, and keyboard bindings in preferences as well (Move window to workspace above/down etc.).

Found this on Wikipedia as an example:

![](https://upload.wikimedia.org/wikipedia/commons/thumb/3/31/Great_Isaiah_Scroll.jpg/800px-Great_Isaiah_Scroll.jpg)

**Reply by smichel17** [reply]
One awesome feature would be if opening the overview when using PaperWM "zoomed out" and displayed all your open windows at once in a view like this image. Currently I feel that the overview is not particularly useful in PaperWM, since it mixes up the layout of windows that I have in my head. The switcher that is shown when moving left/right between windows is actually more useful to me; I seldom open the overview except to launch applications or select a scratch window.

…unfortunately, it seems likely that this would break compatibility with the Vertical Overview extension ([here](https://github.com/RensAlthuis/vertical-overview), since apparently it hasn't been linked in this thread yet). I imagine it's unavoidable that this will happen eventually, though.

**Comment by smichel17** [comment]
I agree that vertical workspaces are integral to how PaperWM works. On the other hand, if PaperWM is working with Vertical Workspaces installed separately, maybe it is best to leave the projects separate and just say PaperWM requires Vertical Workspaces to be installed. That way bugs don't need to be fixed twice, etc. So, I would say don't add it until there is a problem which can only be fixed that way.

…I'm assuming that "module" here is some gnome-shell thing. If you're proposing adding it as a git submodule or something like that and automatically installing it when PaperWM is installed, that seems reasonable to me.

**Reply by michaelhthomas** [reply]
Agreed on both fronts. I think if it's possible to include the extension as a submodule (I've never seen it done, so I'm not entirely certain it's possible) we should. Otherwise, recommending installing it in tandem is definitely not the worst solution.

**Reply by smichel17** [reply]
A submodule is a git feature that lets you say "this folder in the repo is synced to a different upstream git repository". Instead of storing the contents of that folder in the repo, it just stores the `folder name`, `remote url` and `commit hash`. When you run `git submodule init`, it essentially runs `git clone ${remote url} ${folder name} && cd ${folder name} && ${git checkout ${commit hash}` for each submodule.

The downside is that `git submodule init` (and `git submodule update`, the equivalent of `git pull`) don't happen automatically when you `git clone` or `pull`, so if you forget to do them, you end up with the files in your submodule folder being outdated or missing. I *guess* this confusing behavior is because submodules were designed to replace monorepos, and if you have a ton of projects in submodules, you may want to choose just a subset to download after you clone the parent repo.

Anyway, the point is that git submodules are just a way to get the files for vertical-overview into a specific folder; any sort of "automatically activate vertical-overview when PaperWM is activated" feature would have to be included separately.

There's also a less well-known git feature called subtrees, which I think is more like, "I'd like this subfolder to be included in this repo *and* an external repo" (commits to it would appear in *both* repositories). It has a different set of trade-offs. I haven't looked into it enough to fully understand what those are.

**Comment by smichel17** [comment]
Since this was posted, there were a few months where I was using MacOS, which also has horizontal workspaces, similar to Gnome 40+. During this time, it occurred to me that if you mostly fullscreen or split-screen your windows, horizontal workspaces are basically a less-flexible approximation of PaperWM's interface on a single workspace. Especially so with MacOS's implementation, where fullscreen or split-screen windows are automatically moved to their own workspace.

This reinforces to me that vertical workspaces are core to PaperWM; there's no need to duplicate the feature set with both scrollable tiling *and* horizontally scrollable workspaces.

**Reply by paradajz** [reply]
> Since this was posted, there were a few months where I was using MacOS, which also has horizontal workspaces, similar to Gnome 40+. During this time, it occurred to me that if you mostly fullscreen or split-screen your windows, horizontal workspaces are basically a less-flexible approximation of PaperWM's interface on a single workspace.

I came to a similar conclusion, however this can only work on small screen size (ie. laptop). It breaks apart very quickly on large monitors, unless running everything in full screen is your thing. It even works the same way in Windows 10+.

Anyways using PaperWM for few years now makes any other operating system / DE unusable for me.
